addons/delivery/tests/test_delivery_cost.py

```diff
@@ -353,6 +353,70 @@ def test_delivery_cost_gift_card(self):
         self.assertEqual(len(sale_normal_delivery_charges.order_line), 3)
         self.assertEqual(sale_normal_delivery_charges.amount_untaxed, 80.0, "Delivery cost is not Added")
 
+    def test_delivery_real_cost(self):
+        # Set up the carrier
+        product_delivery = self.env['product.product'].create({
+            'name': 'Delivery Charges',
+            'type': 'service',
+            'list_price': 40.0,
+            'categ_id': self.env.ref('delivery.product_category_deliveries').id,
+        })
+        delivery_carrier = self.env['delivery.carrier'].create({
+            'name': 'Delivery Now Free Over 100',
+            'fixed_price': 40,
+            'delivery_type': 'fixed',
+            'invoice_policy': 'real',
+            'product_id': product_delivery.id,
+            'free_over': False,
+        })
+
+        so = self.SaleOrder.create({
+            'partner_id': self.partner_18.id,
+            'partner_invoice_id': self.partner_18.id,
+            'partner_shipping_id': self.partner_18.id,
+            'pricelist_id': self.pricelist.id,
+            'order_line': [(0, 0, {
+                'name': 'PC Assamble + 2GB RAM',
+                'product_id': self.product_4.id,
+                'product_uom_qty': 2,
+                'product_uom': self.product_uom_unit.id,
+                'price_unit': 120.00,
+            })],
+        })
+
+        delivery_wizard = Form(self.env['choose.delivery.carrier'].with_context({
+            'default_order_id': so.id,
+            'default_carrier_id': delivery_carrier.id
+        }))
+        delivery_wizard.save().button_confirm()
+
+        delivery_line = so.order_line.filtered(lambda line: line.is_delivery)
+        self.assertEqual(len(delivery_line), 1, "The delivery line is not present on the SO.")
+        self.assertEqual(delivery_line.price_unit, 0, "The delivery cost should be 0.")
```

Comment:
```suggestion
        self.assertEqual(len(delivery_line), 1)
        self.assertEqual(delivery_line.price_unit, 0, "The invoicing policy of the carrier is set to 'real cost' and that cost is not yet known, hence the 0 value")
```
An assert message should only be written when we need to explain the expected value, not why the assert fails
I mean
```py
self.assertTrue(a, "Should be true")
```
It does not bring anything, the line is clear enough
(You can actually remove most of the below assert messages)